agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/utils/SanitizationHelper.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * ApplicationInsights-Java
+ * Copyright (c) Microsoft Corporation
+ * All rights reserved.
+ *
+ * MIT License
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this
+ * software and associated documentation files (the ""Software""), to deal in the Software
+ * without restriction, including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ * persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.agent.internal.exporter.utils;
+
+import com.microsoft.applicationinsights.agent.internal.common.Strings;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SanitizationHelper {
+  public static final int MAX_KEY_LENGTH = 150;
+  public static final int MAX_VALUE_LENGTH = 8192;
+  public static final int MAX_NAME_LENGTH = 1024;
+  public static final int MAX_ID_LENGTH = 512;
+  public static final int MAX_MESSAGE_LENGTH = 32768;
+  public static final int MAX_URL_LENGTH = 2048;
+
+  /** Function to sanitize both key and value in properties. */
+  @SuppressWarnings("ReturnsNullCollection")
+  public static Map<String, String> sanitizeProperties(Map<String, String> properties) {
+    if (properties == null) {
+      return null;
+    }
+    if (!needsSanitizing(properties)) {
```

Comment:
is it better to avoid this iteration? line 47 iteration, we can check if the size has been exceeded, if yes, sanitize.. otherwise go to next element in the list.. i think this is not necessary.

Reply:
as Trask mention this optimization is to avoid extra memory allocation when properties are valid

Reply:
when it's invalid, two iterations.. i am saying for both valid and invalid cases, one iteration is sufficient.

Reply:
can you show us what you are thinking? I did not see a way to do it in one iteration without memory allocations in the normal case